File: test/four/app/src/main/java/com/z3d/scene/ConstrainSceneView.java
package com.z3d.scene;
import android.content.Context;
import android.opengl.GLES20;
import android.opengl.GLSurfaceView;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import androidx.constraintlayout.solver.widgets.Rectangle;

import com.z3d.base.CallBackFun;
import com.z3d.base.GroupBackFun;
import com.z3d.base.GroupItem;
import com.z3d.display.BuildDisplay3DSprite;
import com.z3d.display.line.GridLineSprite;
import com.z3d.display.particle.CombineParticle;
import com.z3d.display.role.Display3dMovie;
import com.z3d.display.role.SceneChar;
import com.z3d.filemodel.ParticleManager;
import com.z3d.frame3d.Frame3dSprite;
import com.z3d.md5.Md5MoveSprite;
import com.z3d.res.BaseRes;
import com.z3d.res.GroupRes;
import com.z3d.res.SceneRes;
import com.z3d.vo.Vector2D;
import com.z3d.vo.Vector3D;
import org.json.JSONArray;
import org.json.JSONObject;
import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;
public class ConstrainSceneView extends ViewGroup {
    private CallBackFun _initCompleteFun;
    public ConstrainSceneView(Context context,CallBackFun backFun) {
        super(context);
        _initCompleteFun =backFun;
        initData(1);
    }
    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int childCount = getChildCount();
        for (int i = 0; i < childCount; i++) {
            View childView = getChildAt(i);
            childView.layout(i * getWidth(), t, (i + 1) * getWidth(), b);
        }
    }
    public Scene3D _scene3D;
    private void initData( float skipnum){
        _scene3D =new Scene3D();
        GLSurfaceView mGLView=new GLSurfaceView(this.getContext());
        this.addView(mGLView);

        mGLView.setEGLContextClientVersion(2);
        mGLView.setRenderer(new GLSurfaceView.Renderer() {
            @Override
            public void onSurfaceCreated(GL10 gl, EGLConfig config) {
                _scene3D.initData();
                GridLineSprite dis=new GridLineSprite(_scene3D);
                _scene3D.camera3D.distance=300;
                _scene3D.addDisplay(dis);

                _initCompleteFun.StateChange(true);
            }
            @Override
            public void onSurfaceChanged(GL10 gl, int width, int height) {
                GLES20.glViewport(0, 0, width, height);
                _scene3D.camera3D.fovw = width;
                _scene3D.camera3D.fovh = height;
                _scene3D.resizeScene();
            }
            @Override
            public void onDrawFrame(GL10 gl) {
                GLES20.glClearColor(0.0f, 0.0f, 0.0f, 0.2f);
                GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT | GLES20.GL_DEPTH_BUFFER_BIT);
                _scene3D.upFrame();
                _scene3D.camera3D.rotationY+=skipnum;
            }
        });
        mGLView.setRenderMode(GLSurfaceView.RENDERMODE_CONTINUOUSLY);
    }

    public void loadSceneByUrl( String val){
        SceneRes _sceneRes = new SceneRes(_scene3D);
        _sceneRes.load("map/"+val+".txt", new CallBackFun() {
                    @Override
                    public void StateChange(boolean State) {
                        makeOBjData(_scene3D,_sceneRes);
                    }
                }
        );
    }
    private void makeOBjData(Scene3D scene3D,  SceneRes _sceneRes)
    {
        try {
            JSONArray buildItem=     _sceneRes.sceneData.getJSONArray("buildItem");
            for(int i=0;i<buildItem.length();i++){
                this.parsingBuildItem(scene3D,(JSONObject)buildItem.get(i));
            }

            JSONObject fogColor=   _sceneRes.sceneData.getJSONObject("fogColor");
            float d= (float)  _sceneRes.sceneData.getDouble("fogDistance");
            float s=  (float)   _sceneRes.sceneData.getDouble("fogAttenuation");
            scene3D.fogColor=new Vector3D();
            scene3D.fogColor.x=fogColor.getLong("x")/255.0f;
            scene3D.fogColor.y=fogColor.getLong("y")/255.0f;
            scene3D.fogColor.z=fogColor.getLong("z")/255.0f;
            scene3D.fogData=new Vector2D(d * s,1 / ((1 - s) * d));


        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private  void parsingBuildItem(Scene3D scene3D, JSONObject obj)
    {
        try {
            int type = obj.getInt("type");
            if(type== BaseRes.PREFAB_TYPE){
                int id=obj.getInt("id");
                if( id==2){

                }
                BuildDisplay3DSprite tempDis=new BuildDisplay3DSprite();
                tempDis.scene3D =scene3D;
                tempDis.setInfo(obj);
                scene3D.addDisplay(tempDis);
            }
            if(type==BaseRes.SCENE_PARTICLE_TYPE){
                CombineParticle particle = this.getParticleSprite(scene3D,obj);
                scene3D.particleManager.addParticle(particle);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    protected  CombineParticle getParticleSprite(Scene3D scene3D, JSONObject itemObj) {

        try {
            CombineParticle particle =     scene3D.particleManager.getParticleByte(   itemObj.getString( "url"));
            particle.type=0;
            particle.setX((float)itemObj.getDouble("x"));
            particle.setY((float) itemObj.getDouble("y"));
            particle.setZ((float)itemObj.getDouble("z"));
            particle.setScaleX((float)itemObj.getInt("scaleX"));
            particle.setScaleY( (float)itemObj.getInt("scaleY"));
            particle.setScaleZ((float)itemObj.getInt("scaleZ"));
            particle.setRotationX((float)itemObj.getDouble("rotationX"));
            particle.setRotationY((float) itemObj.getDouble("rotationY"));
            particle.setRotationZ((float)itemObj.getDouble("rotationZ"));
            return  particle;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }

    }
    public void  addRoleToSceneByUrl(  String val,Vector3D pos)
    {
        GridLineSprite  dic=new GridLineSprite(_scene3D);
        dic.changeColor(new Vector3D(1,1,1,1));
        _scene3D.addDisplay(dic);
        Display3dMovie sc=new Display3dMovie(_scene3D);
        sc.setRoleUrl("role/"+val);
        sc.scaleX=2;
        sc.scaleY=2;
        sc.scaleZ=2;
        sc.x=pos.x;
        sc.y=pos.y;
        sc.z=pos.z;
        _scene3D.addMovieDisplay(sc);
    }
    public void   playLyf(Scene3D scene3D,String url)
    {
        scene3D.groupDataManager.getGroupData(url, new GroupBackFun() {
            @Override
            public void Bfun(GroupRes groupRes) {
                for (int i = 0; i < groupRes.dataAry.size(); i++) {
                    GroupItem item =  groupRes.dataAry.get(i);
                    if (item.types == BaseRes.SCENE_PARTICLE_TYPE) {

                        ParticleManager particleManager= scene3D.particleManager;
                        CombineParticle particle =    scene3D.particleManager.getParticleByte(item.particleUrl);
                        particleManager.addParticle(particle);
                        Log.d("TAG", "Bfun: ");

                    } else {
                        Log.d("播放的不是单纯特效", "Bfun: ");
                    }
                }
            }
        });
    }

    public void setRect(Rectangle rect){
        ViewGroup.LayoutParams layoutParams=this.getLayoutParams();
        layoutParams.width=rect.width;
        layoutParams.height=rect.height;

        this.setLayoutParams(layoutParams);
    }

    public void addLoadFrame3dRes()
    {
        Frame3dSprite frame3dSprite=new Frame3dSprite(_scene3D);

        _scene3D.addDisplay(frame3dSprite);

    }
    public void   playParticle(String url)
    {
        //"model/"+str +"_lyf.txt"
        url="model/"+"10018" +"_lyf.txt";
        _scene3D.groupDataManager.getGroupData(url, new GroupBackFun() {
            @Override
            public void Bfun(GroupRes groupRes) {
                for (int i = 0; i < groupRes.dataAry.size(); i++) {
                    GroupItem item =  groupRes.dataAry.get(i);
                    if (item.types == BaseRes.SCENE_PARTICLE_TYPE) {
                        ParticleManager particleManager= _scene3D.particleManager;
                        CombineParticle particle =      _scene3D.particleManager.getParticleByte(item.particleUrl);
                        particleManager.addParticle(particle);
                        Log.d("TAG", "Bfun: ");

                    } else {
                        Log.d("播放的不是单纯特效", "Bfun: ");
                    }
                }
            }
        });
    }
    public void addLocaMd5(){
        Md5MoveSprite $sc = new Md5MoveSprite(_scene3D);
        https://webpan.oss-cn-shanghai.aliyuncs.com/res/pan/expmd5/shuangdaonv.jpg
        $sc.setMd5url("pan/expmd5/2/body.md5mesh", "pan/expmd5/2/stand.md5anim", "pan/expmd5/shuangdaonv.jpg");

        _scene3D.addDisplay($sc);
    }
    public void addSceneChar(){
        SceneChar sceneChar=new SceneChar(_scene3D);
        sceneChar.setRoleUrl("role/50011.txt");
        sceneChar.play(SceneChar.CharAction_stand);
        _scene3D.addMovieDisplay(sceneChar);
        sceneChar.addPart(SceneChar.WEAPON_PART ,SceneChar.WEAPON_DEFAULT_SLOT,"model/50011.txt" );
        sceneChar.setMountById("5104");
        sceneChar.play(SceneChar.CharAction_walk);

    }

}

File: h5gl/new/src/pan3d/ConstrainSceneView.ts

import DisplayBaseSprite = Pan3d.DisplayBaseSprite;
import Context3D = Pan3d.Context3D;
import Display3dMovie = Pan3d.Display3dMovie;

module Pan3d {
    export class ConstrainSceneView {
        public renderContext: WebGLRenderingContext;
        public canvas3D: HTMLCanvasElement;
        public scene3D: Scene3D;


        constructor(value: HTMLCanvasElement) {
            this.canvas3D = value;
            var gl: any = this.canvas3D.getContext('webgl', { stencil: true, alpha: true, depth: true, antialias: false })
                || this.canvas3D.getContext('experimental-webgl', { stencil: true, alpha: true, depth: true, antialias: false });

            this.renderContext = gl;
            this.scene3D = new Scene3D(this.renderContext);

            this.scene3D.camera3D.distance = 200;
            this.scene3D.camera3D.rotationX = -30;
            this.scene3D.camera3D.rotationY = 45;
            this.addEvents();

        }

        private addEvents(): void {

            GameMouseManager.getInstance().addMouseEvent(this.canvas3D);
            GameMouseManager.getInstance().uiBlankStage.addEventListener(InteractiveEvent.Down, this.onDown, this);
            GameMouseManager.getInstance().uiBlankStage.addEventListener(InteractiveEvent.Move, this.onMove, this);
            GameMouseManager.getInstance().uiBlankStage.addEventListener(InteractiveEvent.Up, this.onUp, this);
            GameMouseManager.getInstance().uiBlankStage.addEventListener(InteractiveEvent.WheelEvent, this.onWheelEvent, this);
        }
        private downPos: Vector2D;
        private lastCame: Camera3D;
        private onWheelEvent(event: InteractiveEvent): void {
            this.scene3D.camera3D.distance -= event.data;
        }
        private onDown(event: InteractiveEvent): void {
            this.downPos = new Vector2D(event.mouseEvent.x, event.mouseEvent.y)
            this.lastCame = new Camera3D();
            this.lastCame.rotationX = this.scene3D.camera3D.rotationX;
            this.lastCame.rotationY = this.scene3D.camera3D.rotationY;
            this.lastCame.distance = this.scene3D.camera3D.distance;
        }
        private onMove(event: InteractiveEvent): void {
            if (this.downPos) {
                this.scene3D.camera3D.rotationY = this.lastCame.rotationY + (this.downPos.x - event.mouseEvent.x);
                this.scene3D.camera3D.rotationX = this.lastCame.rotationX + (this.downPos.y - event.mouseEvent.y);
            }
        }
        private onUp(event: InteractiveEvent): void {
            this.downPos = null;
        }
        public loadSceneByUrl(value: string): void {
            this.scene3D.camera3D.distance = 1500;
            var sceneRes: SceneRes = new SceneRes(this.scene3D);
            //10002
            //2014
            sceneRes.load(value, () => {
                var buildAry: Array<any> = sceneRes.sceneData.buildItem;
                for (var i: number = 0; i < buildAry.length; i++) {
                    var itemObj: any = buildAry[i];
                    if (itemObj.type == BaseRes.PREFAB_TYPE) {

                        if (itemObj.id != 3) {
                            // continue;
                        }
                        var itemDisplay: Display3DSprite = this.getBuildSprite(itemObj);
                        this.scene3D.addDisplay(itemDisplay);
                    } else if (itemObj.type == BaseRes.SCENE_PARTICLE_TYPE) {

                    }
                }
            }, () => {

            }, () => {

            });
        }
        private getBuildSprite(value: any): Display3DSprite {
            var itemDisplay: Display3DSprite = new Display3DSprite(this.scene3D);
            itemDisplay.setObjUrl(value.objsurl);
            itemDisplay.setMaterialUrl(value.materialurl, value.materialInfoArr);

            if (value.lighturl) {
                itemDisplay.setLighturl(value.lighturl);
            }
            itemDisplay.scaleX = value.scaleX;
            itemDisplay.scaleY = value.scaleY;
            itemDisplay.scaleZ = value.scaleZ;
            itemDisplay.x = value.x;
            itemDisplay.y = value.y;
            itemDisplay.z = value.z;
            itemDisplay.rotationX = value.rotationX;
            itemDisplay.rotationY = value.rotationY;
            itemDisplay.rotationZ = value.rotationZ;

            return itemDisplay;
        }
        public resetSize(): void {
            var mixNum: number = Math.min(document.body.clientWidth, document.body.clientHeight);
            var stageWidth: number = mixNum;
            var stageHeight: number = mixNum;
            this.canvas3D.width = stageWidth;
            this.canvas3D.height = stageHeight;
            this.renderContext.viewport(0, 0, stageWidth, stageHeight);
            // this.canvas3D.style.position = "absolute";
            // this.canvas3D.style.left = "0px";
            // this.canvas3D.style.top = "0px";

            /*
            var mixNum:number=400;
            var stageWidth: number = mixNum;
            var stageHeight: number =mixNum;
            this.canvas3D.width = stageWidth;
            this.canvas3D.height = stageHeight;
            this.renderContext.viewport(0, 0, stageWidth, stageHeight);
            */

        }
        public clearAll():void
        {
            console.log("清理");
            this.scene3D.clearAll();

        }
        public upFrame(): void {
            this.scene3D.upFrame();
        }

        public addRoleToSceneByUrl(val: String, pos: Vector3D): SceneChar {
            var sc: SceneChar = new SceneChar(this.scene3D);
            sc.setRoleUrl("role/" + val + ".txt");
            sc.x = pos.x;
            sc.y = pos.y;
            sc.z = pos.z;
            this.scene3D.addMovieDisplay(sc);
            // sc.addPart(SceneChar.WEAPON_PART, SceneChar.WEAPON_DEFAULT_SLOT, "model/50011.txt");
            return sc;

        }
        public playParticle(name: string): void {
            var url: string = "model/" + name + "_lyf.txt";
            this.scene3D.groupDataManager.getGroupData(this.scene3D.fileRoot + url, (groupRes: GroupRes) => {
                for (var i: number = 0; i < groupRes.dataAry.length; i++) {
                    var item: GroupItem = groupRes.dataAry[i];
                    var posV3d: Vector3D;
                    var rotationV3d: Vector3D;
                    var scaleV3d: Vector3D;
                    if (item.isGroup) {
                        posV3d = new Vector3D(item.x, item.y, item.z);
                        rotationV3d = new Vector3D(item.rotationX, item.rotationY, item.rotationZ);
                        scaleV3d = new Vector3D(item.scaleX, item.scaleY, item.scaleZ);
                    }
                    if (item.types == BaseRes.SCENE_PARTICLE_TYPE) {
                        var particle: CombineParticle = this.scene3D.particleManager.getParticleByte(this.scene3D.fileRoot + item.particleUrl);
                        this.scene3D.particleManager.addParticle(particle);
                    }

                }
            })
        }
    }
}